core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala

```diff
@@ -206,15 +206,28 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
         workerEnv.put("PYTHON_WORKER_FACTORY_SECRET", authHelper.secret)
         // This is equivalent to setting the -u flag; we use it because ipython doesn't support -u:
         workerEnv.put("PYTHONUNBUFFERED", "YES")
-        daemon = pb.start()
-
-        val in = new DataInputStream(daemon.getInputStream)
-        try {
-          daemonPort = in.readInt()
-        } catch {
-          case _: EOFException =>
-            throw new SparkException(s"No port number in $daemonModule's stdout")
```

Comment:
Why don't you just fix exception message or adding an error level log? Seems confusing error doesn't justify retrying

Reply:
Adding a retry sometimes makes the task successful.

Reply:
port 0 in daemon.py will automatically allocate available socket. In which case does it fail?

Reply:
In our production environment , the python sub process sometimes will exit from code 139 .
When I debug it and find the python process is dead.
So , the code `in.ReadInt ()` will throw EOFException not related the port .

Reply:
> When I debug it and find the python process is dead.

Do you know why it was dead? We might better have to make daemon more robust rather than just retrying.

Reply:
I'm Sorry! I only know the daemon sometimes dead and exit with code 139.
We use Yarn mode, so the hierarchical relationship looks like:
Yarn container process
          |
          | ----JVM process
                         |
                         | ----- python process

Reply:
This retry mechanism could reduce the fail probability.

Reply:
I think we should figure out the root cause first before considering retry. This fix basically says you cannot trust os

Reply:
I think the root cause of the exit is the robust  problem with the user code.

Reply:
Can you tell me in which case it can fail by user code?

Reply:
We have a lot job with PySpark. We only find a small number of job causes the issue.
A user tries to catch some exception of Python code, and we will test it.

Reply:
@HyukjinKwon If when starting Python process, but exited caused by a unstable factor of system, do we need to make the effort to retry?

Reply:
Yes I also don't see a reason to retry, insofar as I don't see evidence that it would make subsequent attempts succeed.